[FEATURE] Garder la langue de l'utilisateur quand il la change (PIX-1566).

MelanieMEB opened the pull request:
## :unicorn: Problème
- Actuellement, on peut changer la langue de pix en ajoutant `?lang=en` dans l'URL
- Ce changement disparait si l'on recharge l'application

## :robot: Solution
- Ajout d'un attribut Lang aux Users (coté front et back)
- Ajout d'une route pour mémoriser la nouvelle langue de l'utilisateur en cas de changement
- Lorsqu'on charge l'utilisateur, on récupère aussi sa langue pour la mettre en locale
- Lorsqu'on appel l'application avec le queryParams `lang`, on modifie la langue du users

## :rainbow: Remarques
- Le bouton de switch de langue n'est pas prévu dans cette PR
- Pas de sauvegarde de la langue coté Front actuellement

## :100: Pour tester
- Aller sur https://app-pr2126.review.pix.fr/inscription?lang=en
- S'inscrire
- Voir le site en Anglais
- Se déconnecter et se reconnecter, et être toujours en anglais.

MelanieMEB committed:
[API] Add column lang in users table

MelanieMEB committed:
[API] Add route to change lang

- Create route PATCH /users/{id}/lang/{lang} to modify the lang of the user
- Use the repository method `updateUserAttributes`
- Modify User-Serializer to add attributes Lang

MelanieMEB committed:
[FRONT] Add Lang in User Model and route to modify it

- Add attributes Lang to User Model
- Modify user adapter route to call the route PATCH /users/{id}/lang/{lang}

pix-service commented:
I'm deploying this PR to these urls:

- App: https://app-pr2126.review.pix.fr
- Orga: https://orga-pr2126.review.pix.fr
- Certif: https://certif-pr2126.review.pix.fr
- Admin: https://admin-pr2126.review.pix.fr
- API: https://api-pr2126.review.pix.fr/api/

 Please check it out!

MelanieMEB committed:
[FRONT] Modify Application to get User and update Local

- Create method _getUserAndLocal to simplify tests
- When the user is authenticated, we get it and update locale
- When we transition with params, we updated user

MelanieMEB committed:
[API&FRONT] Save lang when user subscribe

It is hard to get the current lang with ember-intl, so we had a simple text in translation to get it

chrc commented on api/tests/unit/domain/usecases/change-lang_test.js:
```diff
@@ -0,0 +1,26 @@
+    sinon.stub(userRepository, 'updateUserAttributes');
+  });
+
+  it('should accept terms of service of pix-certif', async () => {
```
?

celineung commented on api/lib/application/users/index.js:
```diff
@@ -225,6 +225,24 @@ exports.register = async function(server) {
+          assign: 'requestedUserIsAuthenticatedUser',
+        }],
+        handler: userController.changeLang,
+        notes : [
+          '- **Cette route est restreinte aux utilisateurs authentifiés**\n' +
+          '- Sauvegarde le fait que l\'utilisateur a accepté les dernières Conditions Générales d\'Utilisation de Pix App\n' +
+          '- L’id demandé doit correspondre à celui de l’utilisateur authentifié\n' +
+          '- Le contenu de la requête n\'est pas pris en compte.',
+        ],
```
**suggestion**: les `notes` ne correspondent plus à cette route

```suggestion
        notes : [
          '- **Cette route est restreinte aux utilisateurs authentifiés**\n' +
          '- Modifie la langue de l'utilisateur\n' +
          '- L’id demandé doit correspondre à celui de l’utilisateur authentifié\n' +
          '- La lang contient les deux lettres de la langue choisie.',
        ],
```

celineung commented on api/lib/application/users/user-controller.js:
```diff
@@ -79,6 +79,17 @@ module.exports = {
+  async changeLang(request) {
+    const authenticatedUserId = request.auth.credentials.userId;
+    const lang = request.params.lang;
+    const updatedUser = await usecases.changeLang({
```
```suggestion
    const updatedUser = await usecases.changeUserLang({
```

celineung reviewed (commented):
Fonctionnellement OK =)

MathieuGilet commented on api/lib/application/users/index.js:
```diff
@@ -225,6 +225,24 @@ exports.register = async function(server) {
+    {
+      method: 'PATCH',
+      path: '/api/users/{id}/lang/{lang}',
+      config: {
```
Je rajouterai bien une validation Joi pour vérifier les valeurs possibles de lang: 
`
 validate: {
   params: Joi.object({
     lang: Joi.string.valid('fr', 'en'),
  }),
},
`

MelanieMEB replied:
Bonne proposition, j'ai pris !
